bugsnag-android-core/src/main/java/com/bugsnag/android/EventFilenameInfo.kt

```diff
@@ -0,0 +1,165 @@
+    val suffix: String,
+    val errorTypes: Set<ErrorType>
+) {
+
+    /**
+     * Encodes event information into a filename
+     */
+    fun encode(): String {
+        return String.format(
+            Locale.US,
+            "%d_%s_%s_%s_%s.json",
+            timestamp,
+            apiKey,
+            serializeErrorTypeHeader(errorTypes),
+            uuid,
+            suffix
+        )
+    }
+
+    fun isLaunchCrashReport(): Boolean = suffix == STARTUP_CRASH
+
+    internal companion object {
+        private const val STARTUP_CRASH = "startupcrash"
+        private const val NON_JVM_CRASH = "not-jvm"
+
+        /**
+         * Generates a filename for the Event in the format
+         * "[timestamp]_[apiKey]_[errorTypes]_[UUID]_[startupcrash|not-jvm].json"
+         */
+        @JvmOverloads
+        fun fromEvent(
+            obj: Any,
+            uuid: String = UUID.randomUUID().toString(),
+            apiKey: String?,
+            timestamp: Long = System.currentTimeMillis(),
+            config: ImmutableConfig
+        ): EventFilenameInfo {
+            val sanitizedApiKey = when {
+                obj is Event -> obj.apiKey
+                apiKey.isNullOrEmpty() -> config.apiKey
+                else -> apiKey
+            }
+
+            return EventFilenameInfo(
+                sanitizedApiKey,
+                uuid,
+                timestamp,
+                findSuffixForEvent(obj, config),
+                findErrorTypesForEvent(obj, sanitizedApiKey)
+            )
+        }
+
+        /**
+         * Reads event information from a filename.
+         */
+        fun fromFile(file: File, config: ImmutableConfig): EventFilenameInfo {
+            return EventFilenameInfo(
+                findApiKeyInFilename(file, config),
+                "", // ignore UUID field when reading from file as unused
+                -1, // ignore timestamp when reading from file as unused
+                findSuffixInFilename(file),
+                findErrorTypesInFilename(file)
+            )
+        }
+
+        /**
+         * Retrieves the api key encoded in the filename, or an empty string if this information
+         * is not encoded for the given event
+         */
+        private fun findApiKeyInFilename(file: File, config: ImmutableConfig): String {
+            val name = file.name.replace("_startupcrash.json".toRegex(), "")
+            val start = name.indexOf("_") + 1
+            val end = name.indexOf("_", start)
+            val apiKey = if (start == 0 || end == -1 || end <= start) {
+                null
+            } else {
+                name.substring(start, end)
+            }
+            return apiKey ?: config.apiKey
+        }
+
+        /**
+         * Retrieves the error types encoded in the filename, or an empty string if this
+         * information is not encoded for the given event
+         */
+        internal fun findErrorTypesInFilename(eventFile: File): Set<ErrorType> {
```

Comment:
Is this made more accessible so that it can be called from `EventPayload` which doesn't have a config? IMO, keeping this method private to this class and maintaining the pattern is more important than avoiding adding a config field to `EventPayload`.

Reply:
changed visibility to private and added `config` to `event` constructor